### PR TITLE
Fix App show error when open unlink in Desktop App 

### PR DESCRIPTION
**Pull request by hoangzinh**
### Fixed Issues

$ https://github.com/Expensify/App/issues/19681
PROPOSAL: https://github.com/Expensify/App/issues/19681#issuecomment-1565023149

### Tests

**On Web/mWeb**
1. Add a secondary login account, and leave it unverified
2. Sign out and sign in to the secondary login account above
3. Click on Unlink button
4. Open email account, click on 'the validation link here' to unlink the account
5. Verify that the page shows "Secondary login successfull unlinked" message

**On Desktop**
Ensure you logged out Expensify account in Web browser
1. Add a secondary login account, and leave it unverified
2. Sign out and sign in to the secondary login account above
3. Click on Unlink button
4. Open email account in Web browser, click on 'the validation link here' to unlink the account
5. Verify that the link opened in Web browser and there is no prompt to open the link in Desktop
7. Verify that the Web browser shows "Secondary login successfull unlinked" message

**On Native apps**
Ensure you logged out Expensify account in Web browser
1. Add a secondary login account, and leave it unverified
2. Sign out and sign in to the secondary login account above
3. Click on Unlink button
4. Open email account, click on 'the validation link here' to unlink the account
5. Verify that the Native app shows "Secondary login successfull unlinked" message

- [x] Verify that no errors appear in the JS console

### Offline tests

Can not test when offline
### QA Steps

Same as Tests
- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://github.com/Expensify/App/assets/9639873/479b3af3-fe87-4e25-952f-675c3320b468

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://github.com/Expensify/App/assets/9639873/d1e02c5c-b264-4a23-b175-78437c9cf4d0

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://github.com/Expensify/App/assets/9639873/8c78e031-e100-4e60-93e5-60088d39914d

</details>

<details>
<summary>Desktop</summary>

https://github.com/Expensify/App/assets/9639873/324f054a-7ae9-40e3-9faf-2835f604040f

</details>

<details>
<summary>iOS</summary>

https://github.com/Expensify/App/assets/9639873/d9bc94dd-59c9-49bb-a96a-86f13bed8f3c

</details>

<details>
<summary>Android</summary>

https://github.com/Expensify/App/assets/9639873/ec5b46c0-6cc7-4e46-b41d-aabf0c356512

</details>

### CODE REVIEW

**Comment by Santhosh-Sellavel** [comment]
@hoangzinh I find the Desktop steps confusing.

> Verify that the link opened in a Web browser and there is no prompt to open the link in the Desktop

We won't show open in the desktop prompt for the dev environment. Have you checked by removing the restriction does it still work?

File: src/ROUTES.js [hunk 1]
```diff
@@ -187,4 +187,8 @@ export default {
             isSubReportPageRoute: Boolean(lodashGet(pathSegments, 2)),
         };
     },
+
+    ROUTES_REGEX: {
+        UNLINK_LOGIN: /\/u($|(\/\/*))/,
+    },
```
Comment (Santhosh-Sellavel) [inline src/ROUTES.js:193]: Is it okay to be here?  I think its better to be on CONST

Thoughts @francoisl?

Comment (francoisl) [reply src/ROUTES.js:193]: Yeah agreed. Looks like we even already have a `REGEX` object, which has a similar `ROUTES` sub-object [here](https://github.com/Expensify/App/blob/3324ea72cdabc84ba69e01bed8c56208a86b8e1f/src/CONST.js#L1180-L1182). That sounds like a good place to put this.

**Comment by Santhosh-Sellavel** [comment]
Will get to this tomorrow, please get [this](https://github.com/Expensify/App/pull/23738#discussion_r1278134198) done @hoangzinh

File: src/components/DeeplinkWrapper/index.website.js [hunk 2]
```diff
@@ -33,6 +38,14 @@ class DeeplinkWrapper extends PureComponent {
         return !Browser.isMobile() && typeof navigator === 'object' && typeof navigator.userAgent === 'string' && /Mac/i.test(navigator.userAgent) && !/Electron/i.test(navigator.userAgent);
     }
 
+    // Function to detect if current route should open in Web only.
+    isUnsupportedDeeplinkRoute() {
```
Comment (Santhosh-Sellavel) [inline src/components/DeeplinkWrapper/index.website.js:42]: Comment can more user friendly

**Comment by Santhosh-Sellavel** [comment]
Seems to be tested well, After clicking the link we are redirected to the home page with a success message, but the unlink page is shown for a while after opening the link for unlinking is it a problem?

i.e This page is shown for a while
Screenshot 2023-08-02 at 6 29 45 AM

cc: @hoangzinh @francoisl

**Comment by Santhosh-Sellavel** [comment]
@francoisl @NikkiWines What can we do here?

File: src/components/DeeplinkWrapper/index.website.js [hunk 3]
```diff
@@ -33,6 +38,14 @@ class DeeplinkWrapper extends PureComponent {
         return !Browser.isMobile() && typeof navigator === 'object' && typeof navigator.userAgent === 'string' && /Mac/i.test(navigator.userAgent) && !/Electron/i.test(navigator.userAgent);
     }
 
+    // A function to detect if current route should be opened in Web only.
+    isUnsupportedDeeplinkRoute() {
+        return _.some([CONST.REGEX.ROUTES.UNLINK_LOGIN], (unsupportRouteRegex) => {
+            const routeRegex = new RegExp(unsupportRouteRegex);
+            return routeRegex.test(window.location.pathname);
+        });
+    }
```
Comment (Santhosh-Sellavel) [inline src/components/DeeplinkWrapper/index.website.js:47]: The method name looks sufficient. Instead, we could have a comment inside explaining why the unlink login path was added here. Thoughts?

cc: @hoangzinh @francoisl

Comment (francoisl) [reply src/components/DeeplinkWrapper/index.website.js:47]: (Personal preference) I don't mind the comment outside the function, even if the name sounds sufficient. But overall neutral, I'm ok if we keep it or remove it.

However, yes a comment inside explaining _why_ the unlink route is there sounds like a good idea to me!

**Comment by Santhosh-Sellavel** [comment]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [ ] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [ ] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web & Desktop </summary>

https://github.com/Expensify/App/assets/85645967/57d85ae4-5efd-4d3d-8b89-c05d5c8f4f56

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://github.com/Expensify/App/assets/85645967/ff61927f-c018-4adb-87d3-134357a677b6

</details>

<details>
<summary>Mobile Web - Safari</summary>

Simulator Screenshot - iPhone 14 - 2023-08-07 at 23 56 46

</details>

<details>
<summary>iOS</summary>

https://github.com/Expensify/App/assets/85645967/a5623cb3-a986-40c1-b4b7-8f674588608b

</details>

<details>
<summary>Android</summary>

Didn't upload

</details>

**Comment by Santhosh-Sellavel** [comment]
Please resolve conflicts @hoangzinh!

**Comment by Santhosh-Sellavel** [comment]
bump @hoangzinh!

File: src/components/DeeplinkWrapper/index.website.js [hunk 4]
```diff
@@ -18,12 +18,18 @@ function isMacOSWeb() {
...
-        }
+    const isUnsupportedDeeplinkRoute = useMemo(
+        () =>
+            // According to the design, we don't support unlink in Desktop app https://github.com/Expensify/App/issues/19681#issuecomment-1610353099
+            _.some([CONST.REGEX.ROUTES.UNLINK_LOGIN], (unsupportRouteRegex) => {
+                const routeRegex = new RegExp(unsupportRouteRegex);
+                return routeRegex.test(window.location.pathname);
+            }),
+        [],
+    );
```
Comment (Santhosh-Sellavel) [inline src/components/DeeplinkWrapper/index.website.js:29]: Why do we need to memorize this?

Comment (Santhosh-Sellavel) [reply src/components/DeeplinkWrapper/index.website.js:29]: 1 value in array I don't think its expensive. Let's just have a definition and use it.

**Review by Santhosh-Sellavel — APPROVED** [review]
LGTM, thanks!

**Comment by Santhosh-Sellavel** [comment]
@francoisl bump!

**Review by francoisl — APPROVED** [review]
LGTM, thanks for pushing this through to the end.

File: src/pages/UnlinkLoginPage.js [hunk 5]
```diff
@@ -1,31 +1,60 @@
...
+        // eslint-disable-next-line react-hooks/exhaustive-deps
+    }, []);
+
+    useEffect(() => {
+        // Only navigate when the unlink login request is completed
+        if (!prevIsLoading || props.account.isLoading) {
+            return;
+        }
+
+        Navigation.navigate(ROUTES.HOME);
```
Comment (rushatgabhane) [inline src/pages/UnlinkLoginPage.js:48]: Hey, this line caused a bug - https://github.com/Expensify/App/issues/36618

RCA is [here](https://github.com/Expensify/App/issues/36618#issuecomment-1947913057).

**Pull request merged**